crates/hir-ty/src/display.rs

```diff
@@ -479,42 +527,54 @@ impl HirDisplay for ProjectionTy {
         // `<Param as Trait>::Assoc`
         if !f.display_target.is_source_code() {
             if let TyKind::Placeholder(idx) = self_ty.kind(Interner) {
-                let db = f.db;
-                let id = from_placeholder_idx(db, *idx);
-                let generics = generics(db.upcast(), id.parent);
+                let res = f.format_bounds_with(|f| {
```

Comment:
I feel like we might wanna move this down into the ` if !bounds.is_empty() {` branch body? Recursive bounds like these will be the minority and so we prevent triggering a heap alloc in the case where we aren't actually doing the different rendering here (but only realize so within this closure.